ipa-core/src/helpers/hashing.rs

```diff
@@ -74,9 +70,37 @@ where
         sha.update(&buf);
     }
 
-    assert!(!is_empty, "must not provide an empty iterator");
     // compute hash
-    Hash(sha.finalize())
+    (Hash(sha.finalize()), is_empty)
+}
+
+/// Computes Hash of serializable values from an iterator
+///
+/// This version panics if an empty input is provided. This can offer defense-in-depth
+/// by helping to prevent fail-open bugs when the input should never be empty.
+///
+/// ## Panics
+/// Panics when Iterator is empty.
+pub fn compute_non_empty_hash<I, T, S>(input: I) -> Hash
```

Comment:
Is it better to return `Option<Hash>` here and let callers explicitly decide what to do when the input is empty? I feel like treating 0 as a valid hash is dangerous if you accidentally call `compute_hash` because after that all bets are off

Reply:
How often do you need an empty hash?  Or could those potentially-empty cases be amended with once(foo).chain(input) at the caller?

Reply:
I think we just saw our first case where some shards after shuffle ended up with 0 shares. To me, the fact that hash was never computed, feels like something callers need to be aware of.

Reply:
`Option<Hash>` would mean adding an `expect` to most calls, and would also mean doing something like `unwrap_or(SHA256_HASH_OF_EMPTY_MESSAGE)` for the empty-ok case, since we need to send something to the other helper to signify "I have no data".

I'd rather expose an extra function than prepend a dummy value to force a message to be non-empty, which seems like a kludge.